src/validate.ts

```diff
@@ -74,6 +82,49 @@ export function validate(
     return [];
   }
 
+  function validateFunctionNode(node: FunctionInterface) {
+    const func =
+      node.arguments.length === 4 ? node.arguments[3] : node.arguments[2];
+
+    return [
+      // visit all other arguments
+      ...node.arguments
+        .filter((arg) => arg !== func)
+        .flatMap((arg) => validateEachChildRecursive(arg, visit)),
```

Comment:
Why is `validateFunctionNode` responsible for recursively calling `visit`? This should be delegated higher, no?

Reply:
I only want to validate the closure, not the whole new statement. 

Its valid to so something like

```ts
new Function(stack, 'id', { onSuccess: new EventBusSource(bus.resource) }, async () => {

});
```

Might need to refactor step functions for the same later too.

And its also valid to do something like 

```ts
new Function(stack, 'id', { onSuccess: new FunctionSource(new Function(...).resource) }, async () => {

});
```

where the nested function needs to be validated. So we need to ensure all of the args are validated.

The current step function code will run the step function validation over the entire `new StepFunction` including the args and does not validate the args with their own rules. 

Will add tests to prove.

Reply:
See?

```
    + test/test-files/step-function.ts:57:39 - error Functionless(100): Cannot perform arithmetic on variables in Step Function
    +
    + https://functionless.org/docs/error-codes#cannot-perform-arithmetic-on-variables-in-step-function
    +
    + 57     stateMachineName: `aNamedMachine${stack.region + 2}`,
```

```ts
new StepFunction(
  stack,
  "test",
  {
    stateMachineName: `aNamedMachine${stack.region + 2}`,
  },
  () => {}
);
```

Reply:
I don't think this belongs in the function. This is the outer function's responsibility.

```ts
.flatMap((arg) => validateEachChildRecursive(arg, visit)),
```

Reply:
I think we'll end up removing the recursive checker once the rules get more complex, but it works for now. 

(for example, if you need to switch contexts in the step function closure validator, you'll double process everything in the closure for both the new context and the original one with no control over if both contexts are valid).

Not a real example, but lets say the body of an if was different.

```ts
if() { // original rules on the if
    // different rules in the body
}
```

The other option is to track state, but you cannot inject state back into the recursive call, so everything in the `if` body will be seen for both contexts and the original context will not know the nodes are any different.

The exception is to backtrack and look at the parents of the node to determine the context.